FOCUS #223: Charge Subcategory column (#218)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/adjustmenttype.md

Adjustment Type indicates what kind of after-the-fact adjustment the record represents. Adjustment Type is commonly used to identify changes like credits and refunds.

The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and SHOULD be one of the example allowed values or a value of choosing from the vendor.

## Column ID

AdjustmentType

## Display Name

Adjustment Type

## Description

A positive or negative change in cost applied after the original usage or purchase record(s).

## Content Constraints

| Constraint      | Value                                    |
| :-------------- | :--------------------------------------- |
| Column required | True                                     |
| Data type       | String                                   |
| Allows nulls    | True                                     |
| Value format    | list-of-values                           |

> ### Allowed values

Reviewer: ```suggestion
Allowed values:
```

| Value      | Description                                                                                                                                                                   |
|:----------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Refund | Applied by the vendor where a refund is given to the vendor, refund can be multiple types, the refund type should be noted in the column ChargeType, this could be for example a refund for unused reservations, or unused pre-commit amounts.
|Credits  | Applied when vendor has incorrectly billed usage or applied a credit for usage such as a promo credit, tax error, SLA violation, rounding error, usage correction or pricing correction.
| Balance Transfer | Applied to an account where a customer migrates an account from one account to another where the migrated account had a positive balance | Applied where an account had a positive balance that was not consumed in the previous month

> | General Adjustment | Any adjustment the vendor applies to a customer account that does not have a specific type

Reviewer: ```suggestion
| General Adjustment | Any adjustment the vendor applies to a customer account that does not have a specific type. |

```

Adjustment Type indicates what kind of after-the-fact adjustment the record represents. Adjustment Type is commonly used to identify changes like credits and refunds.

The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and SHOULD be one of the example allowed values or a value of choosing from the vendor.

## Column ID

AdjustmentType

## Display Name

Adjustment Type

## Description

A positive or negative change in cost applied after the original usage or purchase record(s).

## Content Constraints

| Constraint      | Value                                    |
| :-------------- | :--------------------------------------- |
| Column required | True                                     |
| Data type       | String                                   |
| Allows nulls    | True                                     |
| Value format    | list-of-values                           |

> Allowed values:

Reviewer: I believe we show this as a level 3 header to match others...
```suggestion
### Allowed values
```

> | Value      | Description                                                                                                                                                                   |

Reviewer: What about credits?

|:----------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Refund | Applied by the vendor where a refund is given to the vendor, refund can be multiple types, the refund type should be noted in the column ChargeType, this could be for example a refund for unused reservations, or unused pre-commit amounts.
| Usage Correction | Applied when vendor has incorrectly billed usage - example, a virtual machine ran for 700 hours but was billed for 730 hours, the adjustment would show usage correction, with comments or notes stating a correction was made in billing for usage of virtual machines due to billing calculation errors.
| SLA Violation | Applied after customers request refunds for SLA violations. SLA Violations are where a consumed service did not meet the vendor specified SLO.

> | Tax error | Applied where a vendor has applied the wrong tax calculations

Reviewer: Aren't these all refunds?

Author: Technically yes, but it would be good to have these values to describe the refund. Simply having "Refund" would not provide enough detail to categorise the refund.

# Adjustment Type

An adjustment is a positive or negative change in cost applied after the original usage or purchase record(s). Adjustments may be related to one or more charges and are identified by the ChargeType column.

Adjustment Type indicates what kind of after-the-fact adjustment the record represents. Adjustment Type is commonly used to identify changes like credits and refunds.

> The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and SHOULD be one of the example Allowed values or a value of choosing from the vendor.

Reviewer: ```suggestion
The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and SHOULD be one of the example allowed values or a value of choosing from the vendor.
```

## Column ID

AdjustmentType

## Display Name

Adjustment Type

## Description

> Indicates the record represents an after-the-fact change to an invoice or cost of usage that already occurred (e.g., refunds, credits).

Reviewer: We usually take the first sentence and remove the column name for descriptions...
```suggestion
A positive or negative change in cost applied after the original usage or purchase record(s).
```

## Content Constraints

| Constraint      | Value                                    |
| :-------------- | :--------------------------------------- |
| Column required | True                                     |
| Data type       | String                                   |
| Allows nulls    | True                                     |
| Value format    | list-of-values                           |

Allowed values:

| Value      | Description                                                                                                                                                                   |

> |:-----------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
> | Adjustment | Any adjustments that are applied after the original usage or purchase record. Adjustments may be related to multiple charges.  Adjustment types can be any of the example values or a value the vendor chooses, but MUST be added if ChargeType is of type Adjustment.
> 
> 
> Allowed Value Type Definition:
> | Value | Description |

Reviewer: We don't need this extra table
```suggestion
```

# adjustment type

> Adjustment Type indicates the record represents a change to your monthly invoice through after-the-fact adustements (e.g. refunds, pricing changes). An adjustment can be positive or negative.
> Adjustment Type can be used by practitioners to assess any impacts to spend from positive or negative adjustments.

Reviewer: Not all organizations have monthly invoices. But given this happens before the invoice, I'm wondering if we need that explicit reference at all.

Should we start with an intro definition to what an "adjustment" is? I pulled in the definition from ChargeType.

Can we be a little more precise with the use case? I feel like there's something better but will suggest something as a starting point. (I also aligned it to how we usually phrase the scenarios.)
```suggestion
An adjustment is a positive or negative change in cost applied after the original usage or purchase record(s). Adjustments may be related to one or more charges and are identified by the ChargeType column.

Adjustment Type indicates what kind of after-the-fact adjustment the record represents. Adjustment Type is commonly used to identify changes like credits and refunds.
```

> The AdjustmentType column MUST be present and MUST NOT be null or empty. This column is of type String and MUST be one of the allowed values.

Reviewer: We need to qualify when it's null and when it's not.
```suggestion
The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and MUST be one of the allowed values.
```

> - Usage Correction

Reviewer: Move all of these to the table below and add definitions.

Did you get these from somewhere? This is a lot more detailed than I was expecting 🙂 

We should add an "Other" option since we likely won't catch everything.

Author: I got these from invoices i have had from Google, AWS and Azure.

Reviewer: Can you add the explicit things you saw and how they're tracked to the supporting content?

- Price Correction
- GoodWill
- SLA Violation
- Balance Transfer

> - General Adjustment

Reviewer: Should we have something so generic? I worry a provider might only use this value to abide by the spec without giving practitioners what they're actually looking for.

Author: I think we should specify these but also allow providers to put in what they want. These are what i have seen in invoices and examples from vendors.

Reviewer: Ah, so was this your suggestion instead of "Other"? If that's the case, I wonder if we should put a statement like "MUST/SHOULD NOT use "General Adjustment" unless a more specific adjustment type does not exist." (Or maybe a better wording, since not logic is difficult to follow.)

Author: I have updated the whole thing and used SHOULD for allowed values examples

- Promo Credit
- Refund - Reservations
- Refund - Savings Plan

> - Refund - General usage

Reviewer: We should only have one of these. We have other columns to indicate whether it's related to a commitment. Also "reservation" and "savings plan" are provider-specific terms.

Author: This was to show a refund could be multiple types of refund. In Azure, reservation refunds show up in reservations transactions and not usage details. So thought we should maybe call out different refund types.

Reviewer: Understood. My point is that when a refund is for a reservation, the provider should indicate that in the CommitmentDiscountId/Name/Type/Plan columns.

Reviewer: We have an underlying goal to make single-purpose columns that are easy to use for summarization and filtering. In this case, we'd use AdjustmentType to identify refunds and then other columns to identify what the refund was for.

- Tax error
- Rounding Error
- Account Closure

> See [Appendix: adjustment type) for details about allowed values and governance criteria for this dimension.

Reviewer: We don't need this.
```suggestion
```

## Column ID

AdjustmentType

## Display Name

> Adjustments

Reviewer: ```suggestion
Adjustment Type
```

# Adjustment Type

An adjustment is a positive or negative change in cost applied after the original usage or purchase record(s). Adjustments may be related to one or more charges and are identified by the ChargeType column.

Adjustment Type indicates what kind of after-the-fact adjustment the record represents. Adjustment Type is commonly used to identify changes like credits and refunds.

The AdjustmentType column MUST be present and MUST NOT be null or empty when ChargeType is "Adjustment". AdjustmentType MUST be null when ChargeType is not "Adjustment". This column is of type String and SHOULD be one of the example Allowed values or a value of choosing from the vendor.

- Usage Correction
- Price Correction
- GoodWill
- SLA Violation
- Balance Transfer
- General Adjustment
- Promo Credit
- Refund
- Tax error
- Rounding Error
- Account Closure

## Column ID

AdjustmentType

## Display Name

Adjustment Type

## Description

> Indicates whether the record represents an after-the-fact change to an invoice or cost of usage that already occurred (e.g., refunds, credits).

Reviewer 2: Might just be a style thing, but I don't think the positioning of "whether" between two alternatives is necessary here. This column indicates charges from retrospective changes, not charges related to discrete usage.

Author: removed "whether"

| :-------------- | :--------------------------------------- |
| Column required | True                                     |
| Data type       | String                                   |
| Allows nulls    | True                                     |
| Value format    | list-of-values                           |

Allowed values:

| Value      | Description                                                                                                                                                                   |
|:-----------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Adjustment | Any adjustments that are applied after the original usage or purchase record. Adjustments may be related to multiple charges.  Adjustment types can be any of the example values or a value the vendor chooses, but MUST be added if ChargeType is of type Adjustment.

- Usage Correction
- Price Correction
- GoodWill
- SLA Violation
- Balance Transfer
- General Adjustment
- Promo Credit
- Refund
- Tax error
- Rounding Error

> - Account Closure

Reviewer: Sharing this based on our discussion. Not intending to change what we have here...
```
- Refund
  - Usage Correction (quantity?)
  - Price Correction
  - Tax error
- Credit
  - Promo Credit
  - SLA Violation
  - Account Closure
  - GoodWill
- General Adjustment
  - Price Correction
  - Rounding Error (should this be separate)
  - Balance Transfer
```

Author: I have chosen to leave the examples as they where just reordered a bit

|

Allowed Value Type Definition:
|

> - Usage Correction | Applied when vendor has incorrectly billed usage - example, Virual machine ran for 700 hours but was billed for 730 hours, the adjustment would show usage correction, with comments or notes stating a correction was made in billing for usage of virtual machines due to billing calculation errors.

Reviewer 3: "virtual"

> - Price Correction | Applied as an adjustment where a vendor has incorrecly applied the price to a service, this could be a negative or positive value.

Reviewer 3: "incorrectly"

- GoodWill | Similar to General Adjustment
- SLA Violation | Applied after customers request refunds for SLA violations. SLA Violations are where a consumed service did not meet the vendor specified SLO.
- Balance Transfer | Applied to an account where a customer migrates an account from one account to another where the migrated account had a positive balance | Applied where an account had a positive balance that was not consumed in the previous month

> - General Adjustment | Any adjustment the vendor applies to a customer account that does not have a specific type or is an adjusment negotiated by the customer with the vendor

Reviewer 3: "adjustment" not "adjusment"
